### ISSUE TITLE
Dependency-Track Jenkins Plugin Handling of Project Deletion in DT Server

### ISSUE THREAD

**Issue by msymons** [DependencyTrack/dependency-track#263]
### Issue Type:
  - [X] defect report
  - [X] enhancement request
  
### Current Behavior:
The Dependency-Track Jenkins Plugin has problems (can be improved) when it comes to handling a project that has been deleted within Dependency-Track server

### Steps to Reproduce & Expected Behaviour
I am merging "steps to reproduce" and "expected behaviour" in an attempt to better tell a story from beginning to end.  Illustrating using just two projects (although I actually have many more in practice).

I hope that I am not "overloading" the issue by bundling too many suggestions/observations in one go.

Dependency-Track Server is set up with 2 working projects  (analysis succesfully published from Jenkins Jobs A and B using CycloneDX BOM).

```
Project A  (projectId = ddbdfbdc-47a5-4752-839f-5fd548159e16)
Project B  (projectId = c051b275-1437-4807-9c19-1fe6df5de4a6)
```
Project B was then deleted within Dependency-Track.

Attempting to build "job B" (a maven job) in Jenkins now results in Failure, with the following logged in the console:

```
17:30:38 [DependencyTrack] Publishing artifact to Dependency-Track
17:30:40 [DependencyTrack] The specified project could not be found
17:30:40 Build step 'Publish BoM to Dependency-Track' changed build result to FAILURE
```
What I suggest here for improvement is... 

1. Include the DT server URL in the "Publishing artifact" event
2. Identify the specified project that could not be found.
3. DT plugin should allow configuration of whether result should be FAILURE or UNSTABLE (default to FAILURE for backwards compatibility).

At this point, The Jenkins job config.xml for Job B still looks like:

```
    <org.jenkinsci.plugins.DependencyTrack.DependencyTrackPublisher plugin="dependency-track@2.0.2">
      <projectId>c051b275-1437-4807-9c19-1fe6df5de4a6</projectId>
      <artifact>target/bom.xml</artifact>
      <artifactType>bom</artifactType>
      <isScanResult>false</isScanResult>
      <synchronous>true</synchronous>
    </org.jenkinsci.plugins.DependencyTrack.DependencyTrackPublisher>
```
ie, because no edits have yet been made to the job it is (correctly) still referencing the DT project B.

I suggest that the config support recording of the projectName, which can then be used in the console output for user-friendliness.

Now, open Jenkins Job B config for edit. Notice that the project picker in DependencyTrackPublisher is now displaying project A, although the underlying job config.xml still contains ProjectID for B.

I think that this behaviour is problematic; an invitation to things going wrong.  If you make a jenkins job config tweak that is **independent** of DT (eg, add email address to "Recipients" in "E-mail Notification") and save then this now updates the config to use Project A (even though this was not explicitly selected).  This can be seen by the immediate change to projectId in the saved config:

```
    <org.jenkinsci.plugins.DependencyTrack.DependencyTrackPublisher plugin="dependency-track@2.0.2">
      <projectId>ddbdfbdc-47a5-4752-839f-5fd548159e16</projectId>
      <artifact>target/bom.xml</artifact>
      <artifactType>bom</artifactType>
      <isScanResult>false</isScanResult>
      <synchronous>true</synchronous>
    </org.jenkinsci.plugins.DependencyTrack.DependencyTrackPublisher>
```

DT server project A would now be expected to get polluted.  In my case, I did not try it...  I did not want a project with 1000 npm components to morph into one with 200 java components!

What I would expect to see is for the Dependency-Track jenkins plugin to display a warning on the job configuration screen when there is a mismatch between what is retrieved from the DT server (project list) and what is already configured for the job.

ie, red text, similar to the error that one might see when (say) specifying invalid GitHub credentials (etc) on the same job config screen.  This might also be another opportunity to use projectName (see suggestion above) to make a more user-friendly error message.

### Environment:
  
  - Dependency-Track Version:  3.4.0
  - Dependency-Track plugin v2.0.2

**Comment by stevespringett** [comment]
projectName and projectVersion are already properties that get saved. For freestyle builds, it forces the use of projectId and will only record that, but for pipeline jobs you get the option to specify either the projectId or the projectName + projectVersion. 

However, projectName + projectVersion currently doesn't support synchronous publishing mode due to API limitations in DT. 

There are no plans on recording the project name and version for freestyle builds when the projectId is specified, as that would lead to synchronization issues and Jenkins not knowing if the name or version gets modified on DT.

Failure and unstable builds will be coming in the next version as part of Risk Gates configuration for synchronous publishing builds. I'm targeting to have that complete by the 20th or sooner in time for an upcoming talk at OWASP AppSec Cali.

**Comment by msymons** [comment]
Thanks for the speedy response.  Could the commit "Added default selection to top of project dropdown" possibly also solve the problem whereby project auto-creation does not work (reported as #232).

wrt projectName, my thinking was that having projectName recorded in the plugin `config.xml` for a freestyle/maven job  might be used for providing improved of user-feedback but not that it would be used for synchronization itself.  ie, "just for info".  A bit like the Jenkins job description field. 

However, on reflection, just including projectID in console output for **all** builds (even those that are succesful) might be useful useful enough for troubleshooting... allowing better comparison between the log for a passing build with that for a build that has just failed.

Thus...

```
11:22:09 [DependencyTrack] Publishing artifact to Dependency-Track https://DT-foo.bar.com
11:22:09 [DependencyTrack] The artifact was successfully published to project c051b275-1437-4807-9c19-1fe6df5de4a6
11:22:19 [DependencyTrack] Polling Dependency-Track for BoM processing status
11:22:30 [DependencyTrack] Polling Dependency-Track for BoM processing status
11:22:40 [DependencyTrack] Polling Dependency-Track for BoM processing status
11:22:40 [DependencyTrack] Processing findings
```

and:

```
17:30:38 [DependencyTrack] Publishing artifact to Dependency-Track https://DT-foo.bar.com
17:30:40 [DependencyTrack] The specified project c051b275-1437-4807-9c19-1fe6df5de4a6 could not be found
```

**Comment by stevespringett** [comment]
No, these are different things. But I think adding the projectId to the console output makes sense.
